Fix reusing connections if the name redirects to the same plugin

s-hertel opened the pull request:
##### SUMMARY
The existing code `self._connection.matches_name([current_connection])` fixed reloading connection plugins excessively if they were already in the redirect chain, but would still reload connections if a name further back in the redirect chain was used since the current plugin won't know of that alias.

For example, assuming there is no legacy override for the local connection, this would load `ansible.builtin.local` twice:
```yaml
- hosts: all
  tasks:
    - command: echo test
      vars:
        ansible_connection: "{{ item }}"
      loop:
        - ansible.builtin.local
        - local
```
This fixes that.

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->

##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here -->
<!--- A step-by-step reproduction of the problem is helpful if there is no related issue -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```paste below

```

s-hertel committed:
Fix reusing connections if the connection name redirects to the same plugin

nitzmahone commented on lib/ansible/executor/task_executor.py:
```diff
@@ -542,7 +542,15 @@ def _execute(self, variables=None):
-                not self._connection.matches_name([current_connection]) or
+                not (
+                    (load_info := self._shared_loader_obj.connection_loader.get_with_context(
+                        self._play_context.connection,
```
shouldn't this be `current_connection` that we just calculated above?

s-hertel replied:
Yeah, spotted that a bit ago too. It's fixed.

s-hertel committed:
oops, check next connection

nitzmahone commented on lib/ansible/executor/task_executor.py:
```diff
@@ -542,7 +542,15 @@ def _execute(self, variables=None):
                 not getattr(self._connection, 'connected', False) or
-                not self._connection.matches_name([current_connection]) or
+                not (
+                    (load_info := self._shared_loader_obj.connection_loader.get_with_context(
```
So this works, but is doing a lot more work than it needs to in the loop case, since all it really needs is to resolve the name in `current_connection` (ie `find_plugin_with_context`) and compare the resolved name to `self._connection`'s resolved name (rather than instantiating and configuring the plugin, only to drop it on the floor). That's somewhat complicated by the loader's resolved name values getting munged by the things that cache them on the plugin instances, since you can't just compare the raw context resolved value with what's stored on the plugin instance anymore. I'm wondering if it makes more sense to just smuggle out the entire load context on the plugin instead of adding more "slightly different version of the name" properties (and possibly hang the "are these two things really the same thing?" utility method there as well).

Also, I think `_shared_loader_obj` isn't necessary- IIRC the `connection_loader` singleton-ish is directly accessible in TE (we really need to go kill off the _shared_loader_obj stuff I think, it's a trip hazard)

s-hertel replied:
Keeping the whole PluginLoadContext would probably eliminate this issue in the future (as it seems like we keep needing more bits of that). The-already-smuggled redirect list was almost what I needed but it didn't necessarily contain the resolved fqcn.

If the solution is just to resolve the plugin again and compare the names though, it seems like we could get rid of everything except _load_name and the resolved name (currently `ansible_name`). The redirect list is just leftover from module_defaults.

s-hertel replied:
The only other two places that use the aliases list are unnecessary because it only needs to be done for legacy plugins.
* [Here](https://github.com/ansible/ansible/blob/devel/lib/ansible/vars/plugins.py#L77), because legacy plugins are autoloaded and run by default unlike everything else.

    I'm trying to fix that in https://github.com/ansible/ansible/pull/78814/files#diff-5a7d013904669bdf978b4f7c79ae091abb92af2f00066e3c9ccf5a1d8fb5b8b9R46 by loading everything except legacy via the enabled vars plugins list, and then only doing the comparison for legacy plugins.

* And [here](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/loader.py#L1146), all() loses the PluginLoadContext and there's no way to differentiate between legacy/builtin from the _load_name - but that could just use the resolved name and then I can delete the `matches_name()` part of it.

nitzmahone replied:
Agreed on all points above but one: I don't think `ansible_name` is helping us in this case either, unless we perform the same "munging" of the raw PluginLoadContext resolved name that's occurring in `PluginLoader._update_object` (since that only runs when we actually create a plugin instance), and that feels "icky" to me. I'm seeing an ever-increasing amount of special casing two-step around the legacy/builtin stuff that probably mostly doesn't need to be there if we can always only store and compare canonical resolved names for cases like this.

The other thing with `ansible_aliases` is that it's incomplete anyway, since it only caches redirects that were traversed during the resolution of the raw name used to create that specific plugin instance. If you resolved the same plugin with five different raw names, you could potentially end up with five different sets of values in each of their aliases lists, depending on what the redirect topology looks like (an admittedly contrived example, but we've already seen plenty of "abuse" of redirects :laughing: ). Keeping that around seems like a future trip hazard for someone thinking it's got the full picture of all aliases when that's not necessarily true. When I first saw reference to that in an earlier version of this PR, I thought it was coming from a docstring value and was like :fearful:, but even the cached load value is still not necessarily usable in all circumstances.

So that still leaves us with the question of what we should store on the plugin instance (and if we have a deficiency in what's available on PluginLoadContext). I'm not really opposed to adding a `resolved_canonical_name` and `is_builtin` or something to PluginLoadContext, and I'm kinda leaning toward just exposing the entire PLC on the plugin itself, though I shudder to think how it might be abused in the future :laughing:

s-hertel replied:
Thanks for mulling it over. :smile:

`ansible_name` shouldn't be doing any munging. It's just `PluginLoadContext._resolved_fqcn`. And if it's not, that's a bug. I mirrored the `PluginLoadContext.resolved_fqcn` property, but you can ignore the aliases as since it doesn't affect the `ansible_name` (the fqcn is always passed to _update_object).

I don't think we need an `is_builtin` attr when PluginLoadContext already has `plugin_resolved_collection`. If it's legacy it's an empty str, if it's builtin it's always `ansible.builtin` and I used this when fixing the consistency for builtin in https://github.com/ansible/ansible/pull/78700/files. I'm not sure what `resolved_canonical_name` would give us that the `resolved_fqcn` doesn't.

The only possible deficiency I can think of is that PluginLoadContext is not used at all by PluginLoader.all() (which is pertinent since a big part of this was to differentiate between buitlin/legacy). I fixed that by checking whether the path was internal, and if the plugin is specifically loaded later the PluginLoadContext._resolved_fqcn should still match the plugin's ansible_name.

> The other thing with ansible_aliases is that it's incomplete anyway...

Yes, totally agree. That's what my comment above was trying to say. If the correct fix is to resolve the plugin again (that seemed expensive to me which is why I didn't think we'd want to do that), I can get rid of it. As well as the vestigial _redirect_list, which had the exact same issue, as we've discussed many times now.

I can roll https://github.com/ansible/ansible/pull/78814 into this PR if you want review in one place, I wasn't sure how much overlap there'd be for fixing the TE bug. But so far I've replaced the other instances of matches_name and aliases there, and fixed the only place I could spot where ansible_name was incorrectly munged https://github.com/ansible/ansible/pull/78814/files#diff-6790274bc6447a6667f1825052d04d79d60c098cc1e03a252372b56bbd5aa146L869-L870.

nitzmahone commented on lib/ansible/executor/task_executor.py:
```diff
@@ -542,7 +542,15 @@ def _execute(self, variables=None):
+                not (
+                    (load_info := self._shared_loader_obj.connection_loader.get_with_context(
+                        current_connection,
+                        self._play_context,
```
If we can switch back to just a `find_plugin_with_context` instead of instantiating, all these args can also go away.

nitzmahone reviewed (changes requested):
I think the new helpers that were added for other things are actually hurting in this case- let's chat about how to get the info that's needed without creating an instance of the plugin...

s-hertel committed:
Just use find_plugin_with_context

Just use connection_loader instead of self._shared_loader_obj.connection_loader

s-hertel commented:
Merged with #78814
